src/sleep/sleep.c

```diff
@@ -239,36 +237,42 @@ static int execute_s2h(usec_t hibernate_delay_sec) {
         if (r < 0)
                 return r;
 
-        r = rtc_read_time(&original_time);
-        if (r < 0)
-                return r;
+        tfd = timerfd_create(CLOCK_BOOTTIME_ALARM, TFD_NONBLOCK|TFD_CLOEXEC);
+        if (tfd < 0)
+                return log_error_errno(errno, "Error creating timerfd");
 
-        wake_time = original_time + DIV_ROUND_UP(hibernate_delay_sec, USEC_PER_SEC);
-        r = rtc_write_wake_alarm(wake_time);
-        if (r < 0)
-                return r;
+        log_debug("Set timerfd wake alarm for %s",
+                  format_timespan(buf, sizeof(hibernate_delay_sec), hibernate_delay_sec, USEC_PER_SEC));
 
-        log_debug("Set RTC wake alarm for %" PRIu64, wake_time);
+        timespec_store(&ts.it_value, hibernate_delay_sec);
+
+        r = timerfd_settime(tfd, 0, &ts, NULL);
+        if (r < 0)
+                return log_error_errno(errno, "Error setting hibernate timer");
 
         r = execute(suspend_modes, suspend_states);
         if (r < 0)
                 return r;
 
-        /* Reset RTC right-away */
-        r = rtc_write_wake_alarm(0);
+        fds.fd = tfd;
```

Comment:
I'd probably initialize the structure right here:

```c
…
fds = (struct pollfd) {
        .fd = tfd,
        .events = POLLIN,
};
r = poll(&fds, 1, 0);
…
```

I think it's nice to keep this close together, and it also has the benefit we don't initialize the .fd to -1 first only to override later